Implement MambaForSequenceClassification

Adibvafa opened the pull request:
# What does this PR do?
Adds the `MambaForSequenceClassification` model based on `MambaModel` backbone.

We recently published EHRMamba, a state-of-the-art foundation model for Electronic Health Records. This model is built on the same architecture and we will release the trained weights using the MambaForSequenceClassification class.
https://vectorinstitute.github.io/EHRMamba

Fixes #30431 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
      https://github.com/huggingface/transformers/issues/30431
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

As discussed in https://github.com/huggingface/transformers/issues/30431, @ArthurZucker could you take a look? 😊

## Notes

This implementation closely follows the GPT2ForSequenceClassification method, with the exception of pooling the last hidden states before passing them to the classifier to improve efficiency.

Adibvafa committed:
Added MambaForSequenceClassification to src/transformers

Adibvafa committed:
Added MambaForSequenceClassification to src/transformers

Adibvafa committed:
Updated docs with MambaForSequenceClassification

Adibvafa committed:
Updated docs with MambaForSequenceClassification

Adibvafa committed:
Added tests for MambaForSequenceClassification.

Adibvafa committed:
Added tests for MambaForSequenceClassification.

Adibvafa committed:
Fixed style errors.

Adibvafa committed:
Fixed style errors.

Adibvafa committed:
Fixed errors with dummy objects.

Adibvafa committed:
Fixed errors with dummy objects.

Adibvafa committed:
Fixed style issues.

Adibvafa committed:
Fixed style issues.

Adibvafa committed:
Fixed style issues with ruff.

Adibvafa committed:
Fixed style issues with ruff.

Adibvafa committed:
Fixed incorrect example in docstring.

Adibvafa committed:
Fixed incorrect example in docstring.

MichaelYang-lyx commented:
what can I say

ArthurZucker commented on src/transformers/models/mamba/modeling_mamba.py:
```diff
@@ -705,3 +731,154 @@ def forward(
+        return MambaSequenceClassifierOutput(
+            loss=loss,
+            logits=pooled_logits,
+            cache_params=mamba_outputs.cache_params,
```
not sure why you would need the cache params when you are not doing decoding?

Adibvafa replied:
You are right, the classifier output won't need to cache params. Just fixed it with a new commit.

ArthurZucker reviewed (commented):
thanks for opening a PR! 🤗

Adibvafa committed:
Removed cache_params from MambaSequenceClassifierOutput

Adibvafa committed:
Removed cache_params from MambaSequenceClassifierOutput

ArthurZucker commented:
Could you rebase on main and make sure the CIs are green! 🤗

Adibvafa committed:
Merge remote-tracking branch 'upstream/main' into main

Adibvafa committed:
Fixed issues with the incompatibale test with MambaMixer initialization.

Adibvafa committed:
Fixed issues with the incompatibale test with MambaMixer initialization.

Adibvafa commented:
> Could you rebase on main and make sure the CIs are green! 🤗

Of course! It should be good to merge now.
There is a failure test for "MobileViTV2ModelTest" which is unrelated to Mamba.

Adibvafa committed:
Merge branch 'main' of github.com:Adibvafa/MambaForSequenceClassification into main

Adibvafa committed:
Merge remote-tracking branch 'upstream/main' into main

Adibvafa committed:
Remove parameter cache.

Adibvafa committed:
Add cache_params and use_cache for MambaForSequenceClassification.

Adibvafa commented:
@ArthurZucker I did some digging on prior decoder model for classification implementations and realized some of them (e.g. gpt2) use caching. It seems the use case is when you want to do classification at different milestones within the sequence, so by caching, you don't need to restart decoding from beginning and can make this process more efficient. It's a rare use case but worth to have.

Adibvafa committed:
Improve code style.

Adibvafa committed:
Merge branch 'main' of github.com:huggingface/transformers into main

ArthurZucker commented:
Okay!

ArthurZucker commented on src/transformers/models/mamba/modeling_mamba.py:
```diff
@@ -715,3 +741,156 @@ def forward(
+
+        self.config = config
+
+    def forward(self, features, **kwargs):
+        """Forward pass."""
+        # Pooling is done by the forward pass in `MambaForSequenceClassification`
+        x = features
+        x = self.dropout(x)
+        x = self.dense(x)
+        x = self.activation(x)
+        x = self.dropout(x)
+
+        return self.out_proj(x)
```
let's avoid single letter variables please

Adibvafa replied:
Got it, fixed.

ArthurZucker commented on src/transformers/models/mamba/modeling_mamba.py:
```diff
@@ -715,3 +741,156 @@ def forward(
         )
+
+
+class MambaClassificationHead(nn.Module):
```
I think this can be copied from somewhere no?

Adibvafa replied:
The implementation is mostly similar to other ClassificationHead classes of other model types but the ones I checked implemented the class individually. Do you have any abstraction in mind that I can use?

ArthurZucker replied:
I don't think we need to implement the class

ArthurZucker replied:
I'd rather we do like LlamaForSequenceClassification

ArthurZucker commented on src/transformers/models/mamba/modeling_mamba.py:
```diff
@@ -715,3 +741,156 @@ def forward(
+        else:
+            batch_size, _ = inputs_embeds.shape[:2]
+
+        assert (
```
we don't use asserts here, let's rather raise an error

Adibvafa replied:
Makes sense, fixed.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31155). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Adibvafa committed:
Fix merge conflicts with new main.

Adibvafa committed:
Improve modeling_mamba.py using PR reviews.

Adibvafa committed:
Add MambaForSequenceClassification to list of available Mamba models for testing.

Adibvafa committed:
Merge branch 'main' of github.com:huggingface/transformers into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
@ArthurZucker I merged new main HuggingFace into my branch. Should be ready to merge.

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
@ArthurZucker I would love a review!

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Update test_modeling_mamba.py

Adibvafa committed:
Merge branch 'main' into main

mohith7548 commented:
hey any updates on this PR?

mohith7548 commented:
For finetuning 

> The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31155). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

The docs show only the PEFT fine-tuning approach. Does this also support full fine-tuning using Trainer?

Thanks!

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
> For finetuning
> 
> > The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31155). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.
> 
> The docs show only the PEFT fine-tuning approach. Does this also support full fine-tuning using Trainer?
> 
> Thanks!

Yes, it should support it.

Adibvafa commented:
> hey any updates on this PR?

This should merge soon after @ArthurZucker final review.

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
@amyeroberts I haven't heard back from @ArthurZucker for a month, is he doing well?

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

ArthurZucker commented:
Hey sorry, slipped through my notifications

ArthurZucker reviewed (commented):
late sorry

Adibvafa committed:
Change classification head to a linear layer.

Adibvafa commented:
@ArthurZucker Makes sense, I changed the classification head to a linear layer.

Adibvafa committed:
Update mamba configuration by adding use_mambapy.

Adibvafa committed:
Fix merge conflicts and remove classifier dropout

mohith7548 commented:
hey, any updates?

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
@ArthurZucker Pending review!

ArthurZucker reviewed (commented):
🤗

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented on src/transformers/models/mamba/modeling_mamba.py:
```diff
@@ -769,3 +795,130 @@ def forward(
+    """,
+    MAMBA_START_DOCSTRING,
+)
+class MambaForSequenceClassification(MambaPreTrainedModel):
```
I'm not sure if I understand your comment. The forward method of Mamba and Llama for sequence classification seem different. Could you please elaborate? 🤗

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
@ArthurZucker Now that the https://github.com/huggingface/transformers/pull/32080 is merged, can we do a final review for this one too? Also, I would like to add Mamba2ForSequenceClassification to this PR as well so we have both Mamba models with classification capabilities. Then I would be able to release the EHRMamba model on HuggingFace.

mohith7548 commented:
Hey, any update on this?

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Add **kwargs to MambaForSequenceClassification.

Adibvafa commented on src/transformers/models/mamba/modeling_mamba.py:
```diff
@@ -769,3 +795,130 @@ def forward(
+        labels: Optional[torch.LongTensor] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        use_cache: Optional[bool] = None,
```
Good catch! I suppose it won't be necessary (as opposed to `MambaForCausalLM`) but having it is good. I will add it in a commit now.

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa committed:
Merge branch 'huggingface:main' into main

mohith7548 commented:
Hey @Adibvafa, I installed your working version of transformers library from `!pip install -qq git+https://github.com/Adibvafa/MambaForSequenceClassification.git`. The tokenizer and the model loads fine, but when I try to finetune using the Huggingface Trainer, I get CUDA error:

```bash
OutOfMemoryError: CUDA out of memory. Tried to allocate 232.00 MiB (GPU 0; 21.99 GiB total capacity; 20.61 GiB already allocated; 39.38 MiB free; 21.22 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

It looks like the implementation is taking up all the available GPU. Can you please check this?

mohith7548 commented:
I install the your forked version of huggingface repo `%pip install -qq git+https://github.com/Adibvafa/MambaForSequenceClassification.git`. I'm trying to fine-tune on a text-classification dataset. I get the error similar to @Jellymoon. I'm trying to understand your code. Can you help what might be the issue here @Adibvafa ?

```python
mlflow.set_experiment(MLFLOW_EXPERIMENT)
MODEL_OUTPUT_DIR = f"{TRAINER_DIR}/{DATASET_BASE_NAME}/{ts}"

training_args = TrainingArguments(
    output_dir=MODEL_OUTPUT_DIR,
    overwrite_output_dir=True,
    learning_rate=LEARNING_RATE,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=BATCH_SIZE,
    num_train_epochs=EPOCHS,
    weight_decay=WEIGHT_DECAY,
    label_smoothing_factor=LABEL_SMOOTHING, # this causes model to be overconfident. keep non-zero value.
    # bf16=True,
    eval_strategy="epoch",
    save_strategy="epoch",
    group_by_length=True, # smart batching. useful for dynamic padding.
    load_best_model_at_end=True,
    disable_tqdm=False,
    save_total_limit=3,
    metric_for_best_model='f1', # must be on of those defined in `compute_metrics` func
    greater_is_better=True,
    # auto_find_batch_size=True, # will only reduce, might not be optimal.
)

data_collator = DataCollatorWithPadding(tokenizer=tokenizer)

trainer = CustomTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_data["train"].select(range(50)),
    eval_dataset=tokenized_data["valid"],
    tokenizer=tokenizer,
    data_collator=data_collator, # aids in dynamic padding.
    compute_metrics=compute_metrics,
)

# trainer.add_callback(EvalTrainDataCallback(trainer)) # had to remove because it was messing with early stopping.
trainer.add_callback(EarlyStoppingCallback(early_stopping_patience=3, early_stopping_threshold=0.001)) 
```

Std err/output:
```
[2024-08-25 06:37:20,144] [INFO] [real_accelerator.py:203:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/usr/bin/ld: cannot find -laio: No such file or directory
collect2: error: ld returned 1 exit status
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.3
 [WARNING]  using untested triton version (2.3.1), only 1.0.0 is known to be compatible
[rank0]:[W reducer.cpp:1389] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters in the forward pass. This flag results in an extra traversal of the autograd graph every iteration,  which can adversely affect performance. If your model indeed never has any unused parameters in the forward pass, consider turning this flag off. Note that this warning may be a false positive if your model has flow control causing later iterations to have unused parameters. (function operator())

TypeError: Unsupported types (<class 'transformers.cache_utils.MambaCache'>) passed to `_pad_across_processes`. Only nested list/tuple/dicts of objects that are valid for `is_torch_tensor` should be passed.
File <command-3628464069596680>, line 10
      7 mlflow.set_tags(tags)
      9 # Start Training
---> 10 trainer.train()
     12 # Save trained model to local dir
     13 PIPELINE_DIR = f'{MODEL_OUTPUT_DIR}/pipeline'
File /databricks/python/lib/python3.11/site-packages/accelerate/utils/operations.py:128, in recursively_apply(func, data, test_type, error_on_other_type, *args, **kwargs)
    126     return func(data, *args, **kwargs)
    127 elif error_on_other_type:
--> 128     raise TypeError(
    129         f"Unsupported types ({type(data)}) passed to `{func.__name__}`. Only nested list/tuple/dicts of "
    130         f"objects that are valid for `{test_type.__name__}` should be passed."
    131     )
    132 return data
```

mohith7548 commented:
Hey @Jellymoon, the Mamba model, works as expected during the training loop. However, it fails during the evaluation loop. So, I found that it is necessary to set `use_cache=False` when loading the model so that evaluation does not fail.

cc: @Adibvafa 

```python3
model = MambaForSequenceClassification.from_pretrained(
    model_path, 
    num_labels=len(id2label), 
    id2label=id2label, 
    label2id=label2id,
    use_cache=False  # This needs to be passed when using eval and training Mamba for sequence classification otherwise it will raise an error
)
```

mohith7548 commented:
I noticed that the training speed (fine-tuning) is very slow compared to the other HF transformer models. Can something be improved here?

mohith7548 reviewed (approved):
Ran locally. Loaded, Finetuned and did batch inference. Works as expected.

Adibvafa committed:
Merge branch 'huggingface:main' into main

Adibvafa commented:
> @Adibvafa have you tried running this? Whenever the model gets to an evaluation step I get the error below. The code I tried was the huggingface sequence classification tutorial ([link to tutorial](https://huggingface.co/docs/transformers/en/tasks/sequence_classification)) but i used a gpu, replaced "distilbert/distilbert-base-uncased" with "state-spaces/mamba-130m-hf" and i replaced my local modeling_mamba.py with yours so it does load the model.
> 
> ERROR:
> 
> ```
> Traceback (most recent call last):
>   File ".../tutorial_script.py", line 71, in <module>
>   File ".../lib/python3.11/site-packages/transformers/trainer.py", line 3754, in predict
>     output = eval_loop(
>              ^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/transformers/trainer.py", line 3887, in evaluation_loop
>     logits = self.accelerator.pad_across_processes(logits, dim=1, pad_index=-100)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/accelerate/accelerator.py", line 2508, in pad_across_processes
>     return pad_across_processes(tensor, dim=dim, pad_index=pad_index, pad_first=pad_first)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/accelerate/utils/operations.py", line 411, in wrapper
>     return function(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/accelerate/utils/operations.py", line 678, in pad_across_processes
>     return recursively_apply(
>            ^^^^^^^^^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/accelerate/utils/operations.py", line 107, in recursively_apply
>     return honor_type(
>            ^^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/accelerate/utils/operations.py", line 81, in honor_type
>     return type(obj)(generator)
>            ^^^^^^^^^^^^^^^^^^^^
>   File ".../lib/python3.11/site-packages/accelerate/utils/operations.py", line 110, in <genexpr>
>     recursively_apply(
>   File ".../lib/python3.11/site-packages/accelerate/utils/operations.py", line 128, in recursively_apply
>     raise TypeError(
> TypeError: Unsupported types (<class 'transformers.cache_utils.MambaCache'>) passed to `_pad_across_processes`. 
> Only nested list/tuple/dicts of objects that are valid for `is_torch_tensor` should be passed.
> ```

Please provide the specification of what task you were doing and your library versions. I wasn't able to reproduce the issue on my end.

Adibvafa commented:
> Hey @Jellymoon, the Mamba model, works as expected during the training loop. However, it fails during the evaluation loop. So, I found that it is necessary to set `use_cache=False` when loading the model so that evaluation does not fail.
> 
> cc: @Adibvafa
> 
> ```python
> model = MambaForSequenceClassification.from_pretrained(
>     model_path, 
>     num_labels=len(id2label), 
>     id2label=id2label, 
>     label2id=label2id,
>     use_cache=False  # This needs to be passed when using eval and training Mamba for sequence classification otherwise it will raise an error
> )
> ```

I will take a look. Thank you for bringing this up! @Jellymoon @mohith7548

Adibvafa commented:
> I noticed that the training speed (fine-tuning) is very slow compared to the other HF transformer models. Can something be improved here?

Do you have mamba-ssm installed? Is it slow in the classification or in Mamba in general?

Adibvafa commented:
> @Adibvafa, I have `mamba-ssm` installed. However, I realized that it also need `causal-conv1d>=1.4.0` package train faster. Otherwise it was showing some warning related to conv1d that it's gonna use slow/sequential version. Now that I installed `causal-conv1d>=1.4.0` finetuning works as expected.

Amazing!
There is currently a bug with the slow training path that either breaks in low precision training or uses a huge amount of memory at once. I suggest opening the issue for the memory surge. I have opened the issue and currently working on the low precision training error.

mohith7548 commented:
@Adibvafa, a bug in Mamba? or `transformers`? Can you eloborate?  Please share the link of the issue.

mohith7548 commented:
I successfully ran the Mamba model with the new changes you made to the code. Any chance that this will also support the Mamba2 model?

vasqu commented:
> There is currently a bug with the slow training path that either breaks in low precision training or uses a huge amount of memory at once. I suggest opening the issue for the memory surge. I have opened the issue and currently working on the low precision training error.

I think the low-precision bug possibly refers to #32691. The huge amount of memory in the slow path is to be expected though and is one of the reasons why the kernel exists (i.e. to avoid materializing certain tensors etc). Nothing you can really do about this tbh.
cc @mohith7548

vasqu commented on tests/models/mamba/test_modeling_mamba.py:
```diff
@@ -263,17 +275,24 @@ def prepare_config_and_inputs_for_common(self):
     test_pruning = False
     test_head_masking = False  # Mamba does not have attention heads
+    test_model_parallel = False
+    test_mismatched_shapes = False  # MambaMixer follows a different initialization
```
This seems a bit weird to me :thinking: Disabling the flag shouldn't be needed imo.

Could you add `get_input_embeddings` and `set_input_embeddings` methods for the ForSeqClassification class and see if it fixes those tests?

code30x58 commented:
https://youtu.be/tKCQcM5vfac?feature=shared

On Thu, Sep 5, 2024, 11:18 Anton Vlasjuk ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tests/models/mamba/test_modeling_mamba.py
> <https://github.com/huggingface/transformers/pull/31155#discussion_r1745974912>
> :
>
> >      test_pruning = False
>      test_head_masking = False  # Mamba does not have attention heads
> +    test_model_parallel = False
> +    test_mismatched_shapes = False  # MambaMixer follows a different initialization
>
> This seems a bit weird to me 🤔 Disabling the test_mismatched_shapes flag
> shouldn't be needed imo.
>
> Could you add get_input_embeddings and set_input_embeddings methods for
> the ForSeqClassification class and see if it fixes those tests?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/huggingface/transformers/pull/31155#pullrequestreview-2283780436>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A45VJTJ4RD3QL55CZPIYUF3ZVCN6RAVCNFSM6AAAAABIR73MOGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDEOBTG44DANBTGY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

muaviyaijaz commented:
Hi @Adibvafa,

are the pretrained weights still not available for EHRMamba?

Best,
Muaviya

qgallouedec commented:
closing, see https://github.com/huggingface/transformers/issues/30431#issuecomment-5118499701
